### ISSUE TITLE
types: typing error on `bulkWrite` with `AnyBulkWriteOperation<T>`

### ISSUE THREAD

**Issue by andreialecu** [Automattic/mongoose#11985]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.4.0

### Node.js version

16

### MongoDB server version

5

### Description

1) `bulkWrite` will show a typescript error if passed a variable of type `AnyBulkWriteOperation<ISchema>[]`

### Steps to Reproduce

```ts
function modelBulkTest() {
  interface ISchema {
    _id: string;
    num: number;
  }

  const schema = new Schema({
    _id: { type: String },
    num: Number
  });

  const M = model<ISchema>('Test', schema);

  const ops: AnyBulkWriteOperation<ISchema>[] = [
    {
      updateOne: {
        filter: { _id: 'test' },
        update: {
          $inc: { num: 1 }
        },
        upsert: true
      }
    }
  ];
  M.bulkWrite(ops);
}
```

```
const ops: AnyBulkWriteOperation<ISchema>[]
Argument of type 'AnyBulkWriteOperation<ISchema>[]' is not assignable to parameter of type 'AnyBulkWriteOperation<Document>[]'.
  Type 'AnyBulkWriteOperation<ISchema>' is not assignable to type 'AnyBulkWriteOperation<Document>'.
    Type '{ insertOne: InsertOneModel<ISchema>; }' is not assignable to type 'AnyBulkWriteOperation<Document>'.
      Type '{ insertOne: InsertOneModel<ISchema>; }' is not assignable to type '{ insertOne: InsertOneModel<Document>; }'.
        The types of 'insertOne.document' are incompatible between these types.
          Type 'OptionalId<ISchema>' is not assignable to type 'OptionalId<Document>'.
            Type 'OptionalId<ISchema>' is not assignable to type '{ _id?: ObjectId | undefined; }'.
              Types of property '_id' are incompatible.
                Type 'string | undefined' is not assignable to type 'ObjectId | undefined'.
                  Type 'string' is not assignable to type 'ObjectId'.
```

### Expected Behavior

The test should pass.

**Comment by IslandRhythms** [comment]
Please include your imports on future issue submissions
```typescript
import * as mongoose from 'mongoose';
import * as mongodb from 'mongodb';

function modelBulkTest() {
  interface ISchema {
    _id: string;
    num: number;
  }

  const schema = new mongoose.Schema({
    _id: { type: String },
    num: Number
  });

  const M = mongoose.model<ISchema>('Test', schema);

  const ops: mongodb.AnyBulkWriteOperation<ISchema>[] = [
    {
      updateOne: {
        filter: { _id: 'test' },
        update: {
          $inc: { num: 1 }
        },
        upsert: true
      }
    }
  ];
  M.bulkWrite(ops);
}

modelBulkTest();
```

**Comment by RubenatorX** [comment]
We have this problem as well. What can we provide to help you reproduce? And anything we can do to bandaid the problem?

**Comment by vkarpov15** [comment]
@RubenatorX a standalone a code sample that repros the issue you're seeing. Ideally please open a new issue and follow the issue template, but at least a code sample would be helpful.

**Comment by keyCat** [comment]
We have this issue, as well.
The code sample from the author demonstrates it perfectly, IMO.

I will attempt to describe the reason behind it, and I hope you find it useful:

Type `AnyBulkWriteOperation<TSchema extends Document = Document>` is a generic, however, `.bulkWrite()` method has following overloads:
```ts
interface Model<T, TQueryHelpers = {}, TMethodsAndOverrides = {}, TVirtuals = {}, TSchema = any> extends
  NodeJS.EventEmitter,
  AcceptsDiscriminator,
  IndexManager,
  SessionStarter {
  // ...
  bulkWrite(writes: Array<mongodb.AnyBulkWriteOperation>, options: mongodb.BulkWriteOptions & MongooseBulkWriteOptions, callback: Callback<mongodb.BulkWriteResult>): void;
  bulkWrite(writes: Array<mongodb.AnyBulkWriteOperation>, callback: Callback<mongodb.BulkWriteResult>): void;
  bulkWrite(writes: Array<mongodb.AnyBulkWriteOperation>, options?: mongodb.BulkWriteOptions & MongooseBulkWriteOptions): Promise<mongodb.BulkWriteResult>;
  // ...
```
As you can see, a type is not supplied (and cannot be supplied) to `AnyBulkWriteOperation` generic and it becomes locked to `AnyBulkWriteOperation<Document>` type for each and every model method.

**Comment by vkarpov15** [comment]
@keyCat so the issue you're seeing is that `writes` typing isn't sufficiently strict because `AnyBulkWriteOperation` takes `any`?

**Comment by keyCat** [comment]
@vkarpov15 The issue is that `AnyBulkWriteOperation` treated `AnyBulkWriteOperation<Document>` for each and every model and unable to accept anything, except for `AnyBulkWriteOperation<Document>` or `any`, while it should be able to accept `AnyBulkWriteOperation<TSchema>` by default.

**Comment by vkarpov15** [comment]
Should be fixed by #12167

**Comment by Kezzsim** [comment]
<img width="1203" alt="Screen Shot 2023-03-15 at 4 01 57 PM" src="https://user-images.githubusercontent.com/15878154/225429034-b88e2ecc-e537-4bf5-83bf-333bf600602a.png">
Hi, I just upgraded to 7.0.2 and suddenly I have this problem, code that was working just fine demands to be typed as this MongoDB type, but I've tried many configurations for getting it to realize that the data is already structured correctly and none of them work. Is there a way to bypass this?
I've tried using "as AnyBulkWriteOperation" and "as AnyBulkWriteOperation<any>" it never works
